src/irmin/store_intf.ml

```diff
@@ -371,6 +372,25 @@ module type S = sig
     val shallow : Repo.t -> kinded_hash -> tree
     (** [shallow r h] is the shallow tree object with the hash [h]. No check is
         performed to verify if [h] actually exists in [r]. *)
+
+    (** {1 Proofs} *)
+
+    val produce_proof :
+      repo -> kinded_hash -> (tree -> tree Lwt.t) -> Proof.t Lwt.t
```

Comment:
Let's add to the documentation that the user shouldn't hold on to the trees passed to `f`, nor to the trees derived from it. The risk being memory leaks

Reply:
I think we can also try to explicitly reset the global cache at the end of the computation. That'a bit tricky with nested proofs but that could work. I'm also fine to just state this in the doc.

Reply:
Added a comment about this